### PR TITLE
TechDocs: Support configurable working directory as temp dir

### LINKED ISSUE #3333: TechDocs: Support configurable working directory as temp dir
## Feature Suggestion

There is a `backend.workingDirectory` config option available in `app-config.yaml`. If enabled or set, it tells Scaffolder what path in file system to use for storing files that it needs to do its job. This was implemented by @mfrinnstrom in #3062.

When users are using `techdocs-backend` plugin to prepare and generate their docs, it is currently done using a temporary directory. But we want `techdocs-backend` to use the `backend.workingDirectory` path to fetch and store files in the prepare and generate step.

Note: For the "publish" step of TechDocs, we still want to use the `static` directory inside `techdocs-backend` since it serves a different purpose.

## Possible Implementation

Quite similar to how Scaffolder makes use of `backend.workingDirectory`. See PR #3062 for reference.

## Context

Also see #2998

### PR DESCRIPTION
**Pull request by TejasQ**
## Hey, I just made a Pull Request!
This PR fixes #3333 by using any `backend.workingDirectory` specified in [`app-config.yaml`](https://github.com/backstage/backstage/blob/dc7ac4076a0befce1a752c95de2ec2a84a03b633/app-config.yaml#L42) as a temporary directory for TechDocs before falling back to the OS default.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] ~Added or updated documentation~
- [ ] ~Tests for new functionality and regression tests for bug fixes~
- [ ] ~Screenshots attached (for UI changes)~
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/techdocs-backend/src/DocsBuilder/builder.ts [hunk 1]
```diff
@@ -136,13 +137,18 @@ export class DocsBuilder {
...
         this.entity,
       )}`,
     );
 
     // Create a temporary directory to store the generated files in.
-    const tmpdirPath = os.tmpdir();
+    const config = await loadBackendConfig({
+      argv: process.argv,
+      logger: getRootLogger(),
+    });
```
Comment (OrkoHunter) [inline plugins/techdocs-backend/src/DocsBuilder/builder.ts:149]: The plugin already has a `config` instance provided by the Backstage backend to techdocs backend (inside `createRouter` in `plugins/techdocs-backend/src/service/router.ts`).

I think when an instance of `DocsBuilder` is being created inside the `createRouter`, it should forward the `config` instance as well, which can be used here. :)

File: plugins/techdocs-backend/src/DocsBuilder/builder.ts [hunk 2]
```diff
@@ -75,7 +76,7 @@ export class DocsBuilder {
      */
 
     this.logger.info(
-      `Step 1 of 3: Preparing docs for entity ${serializeEntityRef(
+      `Step 1 of 3: Preparing docs for entity ${stringifyEntityRef(
```
Comment (OrkoHunter) [inline plugins/techdocs-backend/src/DocsBuilder/builder.ts:79]: Nice work on updating these 😁

File: plugins/techdocs-backend/src/DocsBuilder/builder.ts (hunk 1 above)
Comment (Rugvip) [reply plugins/techdocs-backend/src/DocsBuilder/builder.ts:149]: Yep! The isn't exported either so should be np passing in the config instance in `router.ts` without it being a breaking change.

Comment (freben) [reply plugins/techdocs-backend/src/DocsBuilder/builder.ts:149]: Yep so adding it effectively to `DocsBuilderArguments`

**Comment by Rugvip** [comment]
@TejasQ any chance you could have a look at this one so we can :shipit: ? :>

**Comment by TejasQ** [comment]
Sorry for the absence! Will sort it out shortly.

**Review by OrkoHunter — APPROVED** [review]
LGTM!

**Review by Rugvip — APPROVED** [review]
Thank you!

**Comment by CiscoRob** [comment]
@TejasQ why did you choose `backend.workingDirectory` for a configuration used explicitly by techdocs plugin when there is a section dedicated to techdocs configuration? 

Would it not have made more sense to use `techdocs.workingDirectory`?

**Comment by OrkoHunter** [comment]
@CiscoRob `backend.workingDirectory` is used by other plugins as well to store temporary files and process them. It makes sense for techdocs to reuse that property and not ask customers to define this in two places.

**Pull request merged**